### ISSUE TITLE
Using it with async methods

### ISSUE THREAD

**Issue by thiagodp** [GoogleChromeLabs/comlink#423]
I'm trying to use ComLink to call an _async_ method in TypeScript, but I'm getting an error: the `deserialize` declared in `transferHandlers` throws an exception.

My code is as follows:

```typescript
// foo-service.ts
export class FooService {
  public async bar(): Promise< string[] > {
    return [ "foo", "bar" ];
  }
}
```
```javascript
// foo-worker.js
import * as ComLink from "comlink";
import { FooService } from "./foo-service";
ComLink.expose( FooService );
```

```typescript
import * as ComLink from "comlink";

const FooService = ComLink.wrap( new Worker( "foo-worker.js" ) );

const run = async () => {
    let worker = await new FooService();
    const promise = await worker.bar();
    const output = await promise;
    console.log( output );
};

run();
```
Is there a way to make this work? That is, how can I call an _async_ method from `FooService`?

**Comment by surma** [comment]
Ah yes, I can see why you might have expected different behavior.

TL;DR: `await worker.bar()` already invokes the async function and awaited the result. No second await necessary. `promise` should already be your array.

This comes from the fact that `await` is really just sugar over `.then()`. And if the value you return from a `.then()` handler has itself a `.then()` function, that function will automatically get called. This is why promise-chaining works.

Example:
```js
x = await fetch("/")
  .then(r => {
    return new Promise(resolve => resolve(4));
  });
```

`x` would be `4` after this, even though I return another promise in `.then()` and not `4` itself.

Even more obscure example:

```js
x = await fetch("/")
  .then(r => {
    return {
		then: f => f(4)
	};
  });
```

Here I don’t even return a proper promise, just an object with a .then function. Even here `x` would be `4`.

**Comment by surma** [comment]
Closing this as this is WAI. If something is still going wrong, please re-open and put an example somewhere (like glitch.com)

**Comment by thiagodp** [comment]
Thank you, @surma 

I created an example _without_ TypeScript and it worked as expected:
https://glitch.com/~balsam-shell

Since my real code is using Parcel with TS, I put another example on CodeSandBox, but it did not work:
https://codesandbox.io/s/mystifying-dan-467bo

-> TS Error when creating `FooService`: `This expression is not constructable.
  Type 'Remote<unknown>' has no construct signatures`.

**Comment by surma** [comment]
I can’t recreate the error you are seeing on codesandbox, probably I am not good with codesandbox. But my hunch is that parcel is down-compiling to ES5 and that it’s going wrong somewhere.

**Comment by thiagodp** [comment]
Me either (not a codesandbox guy). Same problem without Parcel: 
https://codesandbox.io/s/jolly-bash-94b0n

It appears that the problem is when TypeScript tries to create the exposed class. 
Indeed, I can't see any output that shows the error. That message was just a hint from the editor. The script just doesn't work... :(

![image](https://user-images.githubusercontent.com/2997844/72094362-04f89e00-32f5-11ea-9325-712fb3c72fdc.png)

**Comment by thiagodp** [comment]
_Last thing_: it gives the following error when running it on my (local) machine:

![image](https://user-images.githubusercontent.com/2997844/72096501-0e840500-32f9-11ea-9670-05e368e1ca96.png)

**Comment by thiagodp** [comment]
I think I figured out the problem. `comlink.ts`'s `Remote` lacks the constructor signatures in the last object definition. That is:

```typescript
        (...args: any): void;
        new: (...args: any) => void;
```

Thus, the fixed declaration is:
```typescript
export type Remote<T> =
  (
    T extends (...args: infer R1) => infer R2
      ? (...args: R1) => Promisify<R2>
      : unknown
  ) &
  (
    T extends { new (...args: infer R1): infer R2 }
      ? { new (...args: R1): Promise<Remote<R2>> }
      : unknown
  ) &
  (
    T extends Object
      ? { [K in keyof T]: Remote<T[K]> }
      : unknown
  ) &
  (
    T extends string
      ? Promise<string>
      : unknown
  ) &
  (
    T extends number
      ? Promise<number>
      : unknown
  ) &
  (
    T extends boolean
      ? Promise<boolean>
      : unknown
  ) & {
    (...args: any): void; //  ADDED
    new: (...args: any) => void; // ADDED
    [createEndpoint]: MessagePort;
    [releaseProxy]: () => void;
  };
```

I think I'll submit a PR.

Besides, when using ComLink with TypeScript, `tsconfig.json` may need to have `noImplicityAny` set to `false`.
